### ISSUE TITLE
ReturnCount excludeGuardClauses false positives

### ISSUE THREAD

**Issue by Jyosua** [detekt/detekt#8993]
## Expected Behavior
We have some code that includes guard clauses like this:
```kotlin
private fun foo(bar: Bar, barsMap: Map<String, CorrespondingBar>): Bar? {
  val correspondingObject = barsMap[bar.id]
  if(bar == null) {
    log.warn("bar ${bar.id} was not found in bars map")
    return null
  }

  val type = BarsType.fromCode(correspondingObject.code)
  if(type == null) {
    log.warn("Unexpected code of ${correspondingObject.code} received")
    return null
  }

  return bar.copy(
    type = type
  )
}
```

We have set:

```yaml
style:
  ReturnCount:
    excludeGuardClauses: true
  ThrowsCount:
    excludeGuardClauses: true
```

I'd expect these simple guards to not count against the return count. I'm unsure if it's the initial map retrieval or the logging that causes the false positive, but this appears to be not in the spirit of the rule.

## Observed Behavior
Received detekt error:
```
Function foo has 3 return statements which exceeds the limit of 2. [ReturnCount]
```

## Your Environment

* Version of detekt used: 1.23.8
* Version of Gradle used (if applicable): 2.2.5
* Operating System and version: MacOS 15.5

**Comment by BraisGabin** [comment]
> I'm unsure if it's the initial map retrieval or the logging that causes the false positive

Both. The problem here is how to define guard clauses. For example, I would argue that if you need to log then they are not guard clauses, the function is doing something before returning.

If you don't need the log then probably if you add the return in an Elvis operator the rule will be ok with it.

But again, the problem here is how to define a guard clause. Could you give an implementable definition?

**Comment by Jyosua** [comment]
Well, at a minimum I would say that any statement that simply does an early return after checking a state (nullable, empty, etc.) qualifies as a guard clause. I think that most people would probably agree with that much. Logging disqualifying it or not is definitely subjective -- my personal feelings on that subject are that observability/instrumentation/logging is the one categorical exception I am willing to accept as, in practice, not having this type of logging will make troubleshooting post-production release very difficult.

In this specific example, the only way we could still have the same level of insight in post-analysis would be to add an additional layer of abstraction, returning something like a result class that would then indicate if it was successful or not, the return value if successful, and the type of failure if not, then log the failure.

That level of indirection is somewhat antithetical to clear, maintainable code. It would also only be done because the clear and straightforward solution (log and early return) is complained about by the linter unless expressly suppressed. While you can suppress it, I also don't feel that's good behavior to encourage in routine and it seems like it shouldn't be necessary.

**Comment by BraisGabin** [comment]
> does an early return

What is "an early return"? 3 lines? 4 lines? the `if` one statement and the `return`? This is exactly the problem. It's **really** difficult to define what a guard clause is.

Also, it's impossible to know if a function is logging something or updating something in a data base.

**Comment by Jyosua** [comment]
> Impossible

Surely it should be possible to register calls to a specific interface to ignore...? In this case being the slf4j Logger interface...

**Comment by eygraber** [comment]
I _think_ I wrote something in the past that would make this legal:

```kotlin
val correspondingObject = barsMap[bar.id] ?: return null.also {
    log.warn("bar ${bar.id} was not found in bars map")
}
```

**Comment by Jyosua** [comment]
> I _think_ I wrote something in the past that would make this legal:
> 
> val correspondingObject = barsMap[bar.id] ?: return null.also {
>     log.warn("bar ${bar.id} was not found in bars map")
> }

You are remembering correctly. I finally got a chance to give this a go and it worked. We had one developer that generally prefers the format 

```
val foo = thing[id] ?: run {
  log.warn("blah");
  return null;
}
```

and that still complained, so I was a bit surprised to see this way works. I'm also a bit wary of .also{} usage because it tends to ALWAYS run even when you don't expect it to, but in this case it seems to run only on the righthand statement as I had hoped.

**Comment by BraisGabin** [comment]
`foo ?: return null.also { ... }`, to me, looks *really* bad. I know, it is to make detekt happy but some times I think that it's better to suppress detekt.

`foo ?: run { ...; return null }` is a bit better. We could even change the rule to allow that but that would be detekt saying to the world "This is the right way to write guard caluses" and I'm not sut if that's true. In general I prefer an `if` over a scope function.

But I come back to the same problem, how do we define "guard clause" so this type of checks are allowed but not "real code"?

One example from detekt:

```kotlin
override fun visitPrefixExpression(expression: KtPrefixExpression) {
    super.visitPrefixExpression(expression)

    if (expression.baseExpression == null) return
    val operationToken = expression.operationToken
    if (operationToken != KtTokens.PLUS && operationToken != KtTokens.MINUS) return

    if (expression.node.leaves(forward = false)
            .takeWhile { it is PsiWhiteSpace || it is PsiComment }
            .none { it is PsiWhiteSpace && it.textContains('\n') }
    ) {
        return
    }

    analyze(expression) {
        val parentOrSelf = expression.parentBinaryExpressionOrThis()
        if (parentOrSelf.isUsedAsExpression) return
        if (expression.operationReference.mainReference.resolveToSymbol() != null) return
        val message = "This '${parentOrSelf.text}' is not used"
        report(Finding(Entity.from(expression), message))
    }
}
```

I think that this function clearly breaks `ReturnCount` but If we allow "simple assignations" as "guard clauses" then we end up that this function doesn't broke `ReturnCount`. And, to me, that's a problem because make this rule completely useless.

**Comment by Jyosua** [comment]
> foo ?: return null.also { ... }, to me, looks really bad. I know, it is to make detekt happy but some times I think that it's better to suppress detekt.

I actually agree with you here. I would prefer there be a cleaner option to suppress detekt in these instances that don't require manually tagging each function with the suppression. The easy way would be to completely turn off the ReturnCount option, but there are two issues with this:

1. I think these ReturnCount and ThrowCount rules are actually quite helpful when they exclude guard clauses. It's a good heuristic to have.
2. At the time I wrote this issue, I was working on a team full of pedantic people. They would generally have like 1 or 2 people decide the detekt rules and then try to blindly apply them to all projects. Arguing for changes or exceptions was a large headache, because they essentially do not trust other developers with common sense decisions. As a result, only if you could absolutely pinpoint the issue (like excludeGuardClauses being set to true, for example), could you really get a meaningful change in.

I'm sure I'm not the only person in the world that has ever dealt with the latter issue, but because of that environment, that .also{ ... } method becomes easier to get through code review than putting suppressions everywhere, as much as I dislike it.

With respect to your example from detekt, that `.takeWhile { it is PsiWhiteSpace || it is PsiComment } .none { it is PsiWhiteSpace && it.textContains('\n') }` part is the most egregious to me. I probably wouldn't let that go through a code review, but it is hard for a linter to be able to catch that in a clean way that won't also break on more common-sense usages.

I also don't like that analyze(expression) { ... } block being that complex, but I'd argue it's mostly of that aforementioned takeWhile and that `expression.parentBinaryExpressionOrThis()`, which are clearly doing work. Then again I'm not a fan of Kotlin's `function(parameter) { ... }` syntax to begin with because of the nesting and scoping it introduces...

With respect to my original logging example, I think I mentioned this before, but I would prefer it was possible to register an interface that calls to would not count against, or something along those lines. That way you could expressly mark logging or observability stuff without having to suppress anything.

**Comment by BraisGabin** [comment]
> I'm sure I'm not the only person in the world that has ever dealt with the latter issue, but because of that environment, that .also{ ... } method becomes easier to get through code review than putting suppressions everywhere, as much as I dislike it.

To be honest, this triggers me *a lot*. detekt forcing to write bad code because `@Suppress` is forbidden or something like that. Really triggering.

--- 

I would like to know what other @detekt/maintainers think about this issue in general. How could we improve what a "guard clause" is so things like this doesn't happen or at least we minimize them.

**Comment by schalkms** [comment]
> How could we improve what a "guard clause" is so things like this doesn't happen or at least we minimize them

I think describing the behaviour of what detekt detects as a guard clause and maybe more importantly what not could help. Tests might help to list examples for the expected behaviour.

**Comment by Phoenix1808** [comment]
Hi 👋 I dug into this and the root cause looks narrower (and more fixable) than the thread suggests. Sharing my analysis plus a concrete definition, and I'd love direction before opening a PR.

## Root cause

The guard-skipping logic in `yieldStatementsSkippingGuardClauses` (`Junk.kt`) only skips guard clauses that appear *before the first non-guard statement*:

```kotlin
var firstNonGuardFound = false
bodyBlockExpression?.statements?.forEach {
    if (firstNonGuardFound) {
        yield(it)
    } else if (!it.isGuardClause<T>() && !it.isSuperCall() && !it.isScopedAssignment()) {
        firstNonGuardFound = true
        yield(it)
    }
}
```

In the reported `foo`, the first statement is `val correspondingObject = barsMap[bar.id]`. That isn't a guard clause, so `firstNonGuardFound` flips to `true` immediately and every subsequent guard clause gets counted.

Worth highlighting: the logging is **not** the cause. `isIfConditionGuardClause` already accepts a `then` block with statements before the `return` (it only requires the `return` to be `lastBlockStatementOrThis()`), so `if (x == null) { log.warn(...); return null }` is already recognized as a guard *when it appears at the top*. The sole trigger here is the leading `val`:

```kotlin
// flagged today (false positive) — only because of the leading `val`
fun a(x: Int?): Int {
    val y = compute()
    if (x == null) { log("x"); return 0 }
    return y
}

// NOT flagged today — identical guard, just no leading `val`
fun b(x: Int?): Int {
    if (x == null) { log("x"); return 0 }
    val y = compute()
    return y
}
```

## Proposed definition

Keep the existing structural definition of a guard clause, but let the *leading guard section* treat local property declarations as transparent:

> The leading guard section is the maximal prefix of a function's top-level statements consisting only of (a) guard clauses (`if` without `else` ending in `return`/`throw`, or `?:` with a `return`/`throw` RHS), and (b) local `val`/`var` declarations. Any `return`/`throw` after this prefix, or nested inside a non-guard statement, is always counted.

## Re: the `visitPrefixExpression` counter-example

Under this definition the leading section becomes `[guard, val, guard, guard]`, and the `analyze(expression) { ... }` block is the first "real" statement, so the two `return`s inside it stay counted → 2 returns. With `max = 2` it would no longer be flagged.

So this is the trade-off to decide: **with `excludeGuardClauses = true`, should `visitPrefixExpression` still be flagged?** One reading is that its complexity lives in the `analyze {}` logic rather than in scattered guards, so not flagging it is arguably correct under that opt-in. If you'd prefer it stay flagged, a stricter variant is possible (e.g. cap the number of leading guards).

## Unblocking (per @schalkms  suggestion)

Happy to encode the agreed behavior as an IS / IS-NOT table in the rule's tests + KDoc:

**IS a guard clause (excluded):**
- `if (x == null) return`
- `if (x == null) { log(...); return }`
- `value ?: return`
- leading guards interleaved with `val` declarations

**IS NOT (counted):**
- a `return` after the first non-guard / non-`val` statement
- `if (...) return ... else return ...` (has `else`)
- a `return` nested inside a non-guard block

Could you confirm (1) whether allowing `val`/`var` in the leading guard section is acceptable, and (2) the desired verdict on `visitPrefixExpression` under `excludeGuardClauses = true`? Once the definition is agreed I'll open a PR with the implementation + tests. Thanks!

@BraisGabin 
@Jyosua 
@eygraber 
@schalkms

**Comment by Jyosua** [comment]
@Phoenix1808 I hate to be the one to ask this, but would you mind taking the time to format the code in your post using code blocks, etc.? It's really hard to read and understand what you wrote without the proper formatting.

(I understand there's the possibility you wrote this on mobile where it's much harder to do that, but if you've got a minute to use a PC, it'd be greatly appreciated)

**Comment by Jyosua** [comment]
If I'm reading this right, you're saying that:

```kotlin
private fun foo(bar: Bar): Bar? {
  val bar2 = bar.clone()
  if(bar2 == null) {
    log.warn("bar was null")
    return null
  }

  bar2.mutate()
  return bar2
}
```

Would get flagged, but this would not:
```kotlin
private fun foo(bar: Bar): Bar? {
  if(bar == null) {
    log.warn("bar was null")
    return null
  }

  bar.mutate()
  return bar
}
```

Is that right? If that's the case, I think that's fairly problematic for multiple reasons. Off the top of my head, I can think of the following:
- Code where you want to avoid mutating the incoming objects
- Cases where the name you want to use for the function API is different from how you want to name it within the function

The former is a simplistic example, so it's not really a pattern I would follow much.

**Comment by Phoenix1808** [comment]
@Jyosua 

Sorry about the formatting — you're right, the code blocks got flattened. I've edited the original comment above with proper code blocks; hopefully it reads more clearly now. 

On the second point: I think the mix-up is that what you've sketched is the *current* behavior, not my proposal — and it's actually inverted from what I'm suggesting. Two small clarifications first:

- Both of your snippets have only **2** returns, so with the default `max = 2` neither is flagged today regardless — it takes a *third* early return to cross the threshold (the original `foo` in this issue has three, which is why it trips).
- The asymmetry you're probing is real, though: today a leading `val` ends the guard section, so once there are enough guards, the leading-`val` version counts them and trips the limit while the guard-first version doesn't.

My proposal removes exactly that asymmetry: leading `val`/`var` declarations become *transparent* inside the guard section, so the two patterns are treated identically and both pass. Your two reasons — avoiding mutation of the incoming object, and wanting a different local name than the parameter — are exactly the patterns I want this to support, so I think we're aligned on the desired outcome. 

(One caveat: I floated a stricter "`val` initializers must be call-free" variant for the maintainers to consider — but your `val bar2 = bar.clone()` case is a good argument against it, since a `.clone()` call should clearly still be allowed. So I'd lean toward plain `val`/`var` transparency.)

**Comment by Jyosua** [comment]
No no, I'm intentionally describing the current behavior to clarify the current issue. I'm not commenting on your proposal, because while I think it's fine, I suspect the maintainers will have better insight as to the edge cases than I do.

Also opinions on what detekt should or should not police are probably varied -- a determined individual can probably abuse the syntax enough to satisfy detekt and yet do some nasty stuff. I suspect that type of thing would be caught in a code review, but the line in the sand isn't really clear to me as a user of detekt.

**Comment by Phoenix1808** [comment]
> No no, I'm intentionally describing the current behavior to clarify the current issue. I'm not commenting on your proposal, because while I think it's fine, I suspect the maintainers will have better insight as to the edge cases than I do.
> 
> Also opinions on what detekt should or should not police are probably varied -- a determined individual can probably abuse the syntax enough to satisfy detekt and yet do some nasty stuff. I suspect that type of thing would be caught in a code review, but the line in the sand isn't really clear to me as a user of detekt.



> No no, I'm intentionally describing the current behavior to clarify the current issue. I'm not commenting on your proposal, because while I think it's fine, I suspect the maintainers will have better insight as to the edge cases than I do.
> 
> Also opinions on what detekt should or should not police are probably varied -- a determined individual can probably abuse the syntax enough to satisfy detekt and yet do some nasty stuff. I suspect that type of thing would be caught in a code review, but the line in the sand isn't really clear to me as a user of detekt.

Totally agree — I think of this rule as a readability heuristic / nudge rather than something that can be airtight. Anyone determined enough can satisfy any linter while still writing questionable code, and catching that is squarely a code-review concern, not something the rule should try to guarantee.

That's actually the main reason I proposed pinning the behavior down as an explicit IS / IS-NOT table in the tests + KDoc — it won't make the definition perfect, but it makes the "line in the sand" concrete, documented, and reviewable, which is probably the most useful thing we can give users here.

So I'll hold for a maintainer's call on (a) whether leading `val`/`var` transparency is the direction they want and (b) the `visitPrefixExpression` verdict under `excludeGuardClauses = true` — and I'm happy to open a PR with the implementation + that test table as soon as it's confirmed. Thanks for the quick back-and-forth! 🙏

**Comment by schalkms** [comment]
@Phoenix1808 thank you for your time! From my point of view, go ahead with a PR.

Since your posts read for me to be written with the help of AI, please note our [AI contributing guidelines](https://github.com/detekt/detekt/blob/main/.github/CONTRIBUTING.md#a-note-on-ai-generated-contributions).

**Comment by Phoenix1808** [comment]
> [@Phoenix1808](https://github.com/Phoenix1808) thank you for your time! From my point of view, go ahead with a PR.
> 
> Since your posts read for me to be written with the help of AI, please note our [AI contributing guidelines](https://github.com/detekt/detekt/blob/main/.github/CONTRIBUTING.md#a-note-on-ai-generated-contributions).

okkay
